### PR TITLE
fstab-generator: also mount any paths below /usr in initrd

### PR DESCRIPTION
**Pull request by fbuihuu**
For example if /usr/local (where systemd can be installed) is also a mount
point, it should have the same requisite as /usr, ie it should be mounted in
initrd.

### CODE REVIEW

**Comment by bluca** [comment]
You need a rebase to fix the CI errors

**Comment by poettering** [comment]
Uh, is this really desirable? /usr contains the OS, it's absolutely mandatory to be mounted from the initrd. I am not convinced that anything further down below /usr should get the same privilege though. I mean, we have no idea, it could be relevant, could also be some weird shit like /usr/src/ mounted from NFS or so.

hence, unless we are totally and absolutely sure I wouldn't mount anything automatically from the initrd, and keep things we mount there at a minimum. People can use x-systemd.initrd to explicitly mark their mounts if they want to.

**Comment by fbuihuu** [comment]
I agree mounting everything under /usr in initrd is probably be too much.

However I don't see a reason to not do that for /usr/local  especially since /usr/local is not supposed to be mounted over the network (unlike /usr) and when users  build software themselves, /usr/local is usually the default installation path.

So maybe mounting automatically /usr/local only from initrd would be acceptable ?

**Comment by poettering** [comment]
well, /usr/local/ is weird, I am not sure we should do more to support it then the absolute minimum. i.e. as long as you can make it work by adding the mount option i'd not touch it at all.

**Comment by fbuihuu** [comment]
Could you clarify why you consider /usr/local as 'weird' ? I still don't see why we would make it automatically for /usr but not for /usr/local. One could also set `x-systemd.initrd` for /usr.

**Comment by keszybz** [comment]
The difference between `/usr/local` and `/usr` is not in function or requirements, but only in provenance of the files, with `/usr/local` being expressly for the admin. So if we mount `/usr` automatically, I think we should mount `/usr/local` automatically too. In fact, I would lean towards mounting *everything* underneath `/usr` to be mounted, so that whatever partition layout the user has, things just work. This is what this patch currently does. That said, the concern about `/usr/src` and such is valid… Should we mount local filesystems only?

**Comment by fbuihuu** [comment]
Agreed to mount `/usr/local` automatically too since it might contain files involved in the boot process like `/usr` does.

But rather than mounting anything under `/usr` I would instead mount automatically only directories relevant for the boot process. Hence only `/usr`, `/usr/lib*`, `/usr/bin`, `/usr/sbin` and their counterparts in `/usr/local` would be automatically mounted in initrd.

And I wouldn't consider the fact that they're mounted over the network because  `/usr` is always needed very early anyways whereas `/usr/local` is seldomly mounted over the network (however giving the possibility to the admin to disable automatic mounting in initrd might be helpful for corner cases where admins know that `/usr/local` won't contain boot critical applications and `/usr/local` is mounted over the network).

**Comment by fbuihuu** [comment]
@poettering @keszybz I forced pushed a new version that only mount automatically `/usr` and `/usr/local`, see the reasons in the commit message.

PTAL

**Comment by keszybz** [comment]
Yeah, I think this is reasonable. @poettering ?

**Comment by poettering** [comment]
So, we had multiple reports in the past of people putting /usr/local on some NFS or so (to share the locally-built software across a set of machines), and expecting this to work. This always was a mess because NFS of course shows up pretty late usually, but service files and such are read early. It's why I am allergic to /usr/local in general, i.e. we never know when it shows up, there are no guarantess. So far there' no remotely accepted de-facto standard even whether it has to show up by local-fs.target or by remote-fs.target. The NFS people prefer the latter. You'd need the former if you want it to be around for most services at the minimum.

Now you are proposing to move it even earlier than that, into the initrd. Now while this of course would fix one side of the mess (i.e. unit files could suddenly safely be placed in /usr/local/) this neglects the whole history of people actually assuming it's a late-boot thing they can easily provide with traditional NFS that runs as regular system service.

That's why I am saying it's icky. It always was. People don't agree what it should be. Some people even make it a classic autofs mount, assuming this would work, even though it#s likely going to explode.

hence, i'd always stay away from that dir. People don't agree on the semantics, and willy-nilly moving it into the initrd sounds a bit too much, since it means people would also accept to always do NFS in initrd too if needed.

To summarize: the dir is weird. It has no clear semantics and lifecycle guarantees. Changing these now will make some bits better and other worse, and I really doubt the net benefit is worth it. I'd really leave this to people to configure themselves, i.e. figure out whether they want it on early boot, late boot or initrd. 

The fact that it apparently is accepted practice to that /usr/local can be on NFS even though it explicitly has "local" in the name makes things particularly icky to me.

(Oh, and besides that it shouldn't be below /usr in the first place. Sticking /usr/local into /usr is about as smart as sticking /boot/efi into /boot: it makes no sense, the dirs are conceptually not part of each other, hence should not be nested, but accessible independently of each other)

**Comment by poettering** [comment]
(and did i mention that the fact it is a shared namespace with stuff from different origins is really broken. i.e. if you have stuff from multiple sources you must spit of the namespace. /opt/ is a bit better there, even thogh pretty broken, in other ways)

**Comment by poettering** [comment]
(anyway, you asked why i think it's 'weird'... ;-))

**Comment by poettering** [comment]
btw, people putting /usr/local/ on NFS is the topic of many support forum discussions on the network. Yes, people do stuff like this IRL and expect it to work.

https://unix.stackexchange.com/questions/402493/rhel-usr-local-v-s-usr-on-nfs-mount-unmount-order
https://www.linuxquestions.org/questions/linux-networking-3/mounting-usr-local-on-nfs-mount-483001/
https://www.selflinux.org/selflinux/html/nfs03.html
https://www-zeuthen.desy.de/technisches_seminar/texte/gsi_und_sicherheit.pdf

and fuck, ubuntu even suggests to do this in their NFS docs:

https://help.ubuntu.com/community/SettingUpNFSHowTo

and that's just a brief google search...

**Comment by bluca** [comment]
do we have a way to check whether it's remote or local?

**Comment by poettering** [comment]
> do we have a way to check whether it's remote or local?

that wouldn't suffice. the initrd might be able to copy with NFS or not. I mean, dracut at least optionally support nfs root.

our fstab-generator is supposed to be agnostic to initrd implementaitons however...

I think the requirement to add the `x-systemd.initrd` mount option isn't too bad. Splitting the dir out is an exotic thing, people should be able to add the mount optiont too, if they need it.

**Comment by poettering** [comment]
(Also, let's not forget that people traditionally wanted to put /usr on NFS, mounted late, and only our usrmerge efforts put an end to that. If you want to make assumptions about how /usr/local/ is mounted, you basically have to go through the same stuff we did back then if you want to declare that it's one specific way now. You'd have to set up stuff like we did back then: https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/ — but I am not sure /usr/local is really worth salvaging...)

**Comment by fbuihuu** [comment]
> btw, people putting /usr/local/ on NFS is the topic of many support forum discussions on the network. Yes, people do stuff like this IRL and expect it to work.

Yeah it's always possible to find some people doing weird things with google ;)

Does that mean that we should take it for granted ? I'm not sure...

The definition of /usr/local  is however pretty clear: it's the place used by sysadmin to install software *locally*. Therefore putting /usr/local on NFS to share the locally-built software across a set of machines defeat the whole purpose of /usr/local.

But OK I don't know how frequent this use case can be found on real setups.

However I still don't see why we should treat /usr differently though especially since this directory *is* shareable.

**Comment by poettering** [comment]
I don't understand why there's any need to make this automatic? if people want this mounted early, all they have to do is add the mount option. I am not sure what the benefit is of doing anything about this. This fixes no bug after all. All it does is enable one option by default, that people often actually don't want to have as default. I seriously fail to see the big reason to change anything about it.

I mean, there are ugly things we can fix everywere, but this isn#t one of them.

**Comment by poettering** [comment]
> The definition of /usr/local is however pretty clear: it's the place used by sysadmin to install software _locally_. Therefore putting /usr/local on NFS to share the locally-built software across a set of machines defeat the whole purpose of /usr/local.

Wel, people differ in what is "locally", local to this network? local to this organization? local to this system? who knows. I think the only thing people can agree on is that it doesn't come from the OS vendor.

Also, the description you picked also applies to /opt/. So you make it sound as if there was no difference there, but of course there is.

**Comment by fbuihuu** [comment]
> I don't understand why there's any need to make this automatic?

Because we already do so for /usr and /usr/local has (in theory) more reasons to be mounted automatically than /usr has.

**Comment by fbuihuu** [comment]
> > The definition of /usr/local is however pretty clear: it's the place used by sysadmin to install software _locally_. Therefore putting /usr/local on NFS to share the locally-built software across a set of machines defeat the whole purpose of /usr/local.
> 
> Wel, people differ in what is "locally", local to this network? local to this organization? local to this system? who knows. I think the only thing people can agree on is that it doesn't come from the OS vendor.

Maybe it's clearer with: "it's the place used by *the* sysadmin to install software _locally_".

"The sysadmin" (singular) suggests to me that it's local to his system.

It's actually the way it's described in https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.pdf

**Comment by poettering** [comment]
> Because we already do so for /usr and /usr/local has (in theory) more reasons to be mounted automatically than /usr has.

I am sorry? We cannot boot without /usr/. We *must* mount that early. /usr/local/ otoh is optional stuff that is totally not needed for booting. I really don#t follow anymrore?

**Comment by poettering** [comment]
> Maybe it's clearer with: "it's the place used by _the_ sysadmin to install software _locally_".
> 
> "The sysadmin" (singular) suggests to me that it's local to his system.
> 
> It's actually the way it's described in https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.pdf

sysadmins usually care for more than a single system? also, again, IRL people mount /usr/local all the time from NFS; and ubuntu as mentioned even *suggests* doing so in their docs.

**Comment by fbuihuu** [comment]
> I am sorry? We cannot boot without /usr/. We _must_ mount that early. /usr/local/ otoh is optional stuff that is totally not needed for booting. I really don#t follow anymrore?

As soon as /usr/local contains unit files that are required by the initial transaction, /usr/local is not optional anymore.

**Comment by poettering** [comment]
> As soon as /usr/local contains unit files that are required by the initial transaction, /usr/local is not optional anymore.

Yeah, then the person who dropped it there should also update the fstab option for it?

**Comment by fbuihuu** [comment]
Yes I guess that would be OK since the installation of software in /usr/local is supposed to come from the sysadmin.

I'm wondering if there's an appropriate place in the documentation for describing it.

**Comment by poettering** [comment]
Your distro's documentation, if it decides to support /usr/local/ as first-class concept?

**Comment by fbuihuu** [comment]
OK
